js/uid2-sdk-3.0.0.ts

```diff
@@ -117,118 +113,123 @@ export class UID2 {
+        return this._tokenPromiseHandler.createMaybeDeferredPromise(token ?? null);
     }
     public isLoginRequired() {
-        return this.initialised() ? !this._identity : undefined;
+        if (!this._initComplete) return undefined;
+        return !(this.isLoggedIn() || this._apiClient?.hasActiveRequests());
     }
     public disconnect() {
-        this.abort();
-
-        // TODO: This silently fails to clear the cookie if init hasn't been called and a cookieDomain is used
+        this.abort(`UID2 SDK disconnected.`);
+        // Note: This silently fails to clear the cookie if init hasn't been called and a cookieDomain is used!
         if (this._cookieManager) this._cookieManager.removeCookie();
         else new UID2CookieManager({ cookieName: UID2.COOKIE_NAME }).removeCookie();
-
         this._identity = undefined;
-        this._lastStatus = UID2.IdentityStatus.INVALID;
-
-        const promises = this._promises;
-        this._promises = [];
-        promises.forEach(p => p.reject(new Error("disconnect()")));
+        this.runCallbacks(UID2.EventType.IdentityUpdated, { identity: null });
     }
-    public abort() {
-        this._initCalled = true;
+    
+    // Note: This doesn't invoke callbacks. It's a hard, silent reset.
+    public abort(reason?: string) {
+        this._initComplete = true;
+        this._tokenPromiseHandler.rejectAllPromises(reason ?? new Error(`UID2 SDK aborted.`));
         if (typeof this._refreshTimerId !== 'undefined') {
             clearTimeout(this._refreshTimerId);
             this._refreshTimerId = undefined;
         }
         if (this._apiClient) this._apiClient.abortActiveRequests();
     }
 
-    // PRIVATE STATE
-    
-    _initCalled = false;
-    _opts;
-    _identity;
-    _lastStatus;
-    _refreshTimerId;
-    _refreshVersion;
-    _promises: PromiseOutcome<string>[] = [];
-    private _cookieManager: UID2CookieManager;
-    private _apiClient: Uid2ApiClient;
+    private _opts: Uid2Options | null = null;
+    private _identity: Uid2Identity | null | undefined;
+    private _initComplete = false;
 
-    // PRIVATE METHODS
+    private _cookieManager: UID2CookieManager | undefined;
+    private _apiClient: Uid2ApiClient | undefined;
 
-    private initialised() {
-        return typeof this._lastStatus !== 'undefined'; 
-    }
-    private temporarilyUnavailable() {
-        return this._lastStatus === UID2.IdentityStatus.EXPIRED; 
+    private static _sentSdkLoaded = false;
+    private _sentInit = false;    
+    private callbackPushInterceptor(...args: any[]) {
+        const pushResult = Array.prototype.push.apply(this.callbacks, args);
+        for (const c of args) {
+            if (UID2._sentSdkLoaded) this.safeRunCallback(c, EventType.SdkLoaded, { });
+            if (this._sentInit) this.safeRunCallback(c, EventType.InitCompleted, { identity: this._identity ?? null });
+        }
+        UID2._sentSdkLoaded = true;
+        this._sentInit = true;
```

Comment:
everytime a new callback is pushed, we will trigger `this.safeRunCallback` with  `EventType.SdkLoaded` and `EventType.InitCompleted` events?

Reply:
I'm not sure if I've documented that well enough yet - but yes, that's the goal. The idea is that if you provide a callback (whether it's before or after init) you will definitely get SdkLoaded and InitCompleted. If you provide the callback before our SDK loads, there might be a delay between each one. If you provide it after init is complete you still get those events, but they'll happen immediately.